### ISSUE TITLE
hangs，can not continue.

### ISSUE THREAD

**Issue by gaogaoSpark** [huggingface/optimum-benchmark#189]
### pytorch_llama.yaml
```
defaults:
  - experiment # inheriting experiment schema
  - benchmark: inference
  - launcher: torchrun
  - backend: pytorch
  - _self_ # for hydra 1.1 compatibility
  - override hydra/job_logging: colorlog # colorful logging
  - override hydra/hydra_logging: colorlog # colorful logging

experiment_name: pytorch_llama

launcher:
  #device_isolation_action: warn
  device_isolation: true
  nproc_per_node: 4

backend:
  device: cuda
  device_ids: 0,1,2,3
  no_weights: true
  model: FlagAlpha/Llama2-Chinese-13b-Chat
  task: text-generation

benchmark:
  input_shapes:
    batch_size: 12
    sequence_length: 2048
  new_tokens: 256
```
### logs
![log](https://github.com/huggingface/optimum-benchmark/assets/16930471/f699a0d1-c771-4165-a66c-edf88635a804)

**Comment by gaogaoSpark** [comment]
> is this using the current main ? I added many `torch.distributed.barrier()` calls in key points in the benchmark to avoid the hanging. can you please make sure you're on the latest commit, run the benchmark with `LOG_ALL_RANKS=1 optimum-benchmark ...` and report the full log.

1.  I'm sure，current is main
2. I'm sure,hangs at this point
 `optimum_benchmark/launchers/torchrun/launcher.py"`
```
reports = []
while not queue.empty():
    reports.append(queue.get()) # queue can not get value,it wait report

```
3. retry with `LOG_ALL_RANKS=1`
![image](https://github.com/huggingface/optimum-benchmark/assets/16930471/fb96a591-6675-4699-9754-deac0139ae46)

**Comment by IlyasMoutawwakil** [comment]
thanks, that's weird 😅
so I couldn't reproduce it on my machine (classic `torch.distributed` hanging).
here's an attempt of fixing it by using the lock in main process before using queue:
https://github.com/huggingface/optimum-benchmark/pull/191
please report to me if it solves your issue

**Comment by gaogaoSpark** [comment]
Unfortunately， `fix-hanging` still unable to pass.

 I will provide more detailed information below.
add trace code at `optimum_benchmark/launchers/torchrun/launcher.py`

![image](https://github.com/huggingface/optimum-benchmark/assets/16930471/042a0c58-b281-4aeb-845c-47fa54b08b01)

log :
![image](https://github.com/huggingface/optimum-benchmark/assets/16930471/601c3a4d-6a3a-4653-8e48-9f8f6c1cc7fa)

**Comment by IlyasMoutawwakil** [comment]
are you running your code inside a docker container ?
if so you might need to specify some shared memory `--shm-size=1g` 
and inter process communication `--ipc=private`

**Comment by gaogaoSpark** [comment]
running on the physical machine.
Machine info:
```
OS: Ubuntu 20.04.1 LTS
CPU: 128C
Memroy: 1T
Disk: 10T
GPU: NVIDIA A800-SXM4-80GB * 8
Driver： Driver Version: 530.30.02  CUDA Version: 12.1
```

**Comment by IlyasMoutawwakil** [comment]
I came around this issue here https://github.com/pytorch/pytorch/issues/124714#issuecomment-2088046055 which says that using barriers without setting device might be the cause of this sometimes, as the barrier will try to infer the right devices and it can be wrong. I have pushed this change to `fix-hanging`.

**Comment by gaogaoSpark** [comment]
1.We have different types of devices, such as NVIDIA, AMD.The same situation has occurred.
2.In fact, all `inference` tasks have ended,  occurred during the aggregation of reports.
3.Currently, I have developed a small tool that obtains results by parsing logs.Of course, it is necessary to set `LOG-ALL-RANKS=1`
4.I have a suggestion. Simply put, it means that the child process writes the results to a file, and the main process analyzes these files to obtain the final report.

**Comment by IlyasMoutawwakil** [comment]
@gaogaoSpark thanks for the suggestion, let's go with the file system as it seems more fault tolerant 🤗

**Comment by IlyasMoutawwakil** [comment]
fixed in #191

**Comment by IlyasMoutawwakil** [comment]
@gaogaoSpark the file saving method pollutes the working directory and also results in false positives ; imaging you ran a benchmark the first time and then ran a second one in the same directory, which may results in these reports being taken as reports from current benchmark.
tldr; I added some signal handling to torchrun launcher that forces all ranks to exit #193 and would be very grateful if you could test it in your environment.

**Comment by gaogaoSpark** [comment]
Great！the test passed .
![image](https://github.com/huggingface/optimum-benchmark/assets/16930471/db192d6d-41d6-4d2f-93a9-1e8cbebd47d4)
